### PR TITLE
clustermesh-apiserver: ExternalTrafficPolicy and internalTrafficPolicy can now be changed.

### PR DESCRIPTION
**Pull request by kahirokunn**
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

If externalTrafficPolicy is set to Local, the IP address of the connection source can be retained.
This is very important for observer days, so you may want to enable it in many cases.

https://kubernetes.io/docs/tutorials/services/source-ip/#source-ip-for-services-with-type-nodeport

```release-note
clustermesh-apiserver: ExternalTrafficPolicy and internalTrafficPolicy can now be changed.
```

### CODE REVIEW

**Review by tommyp1ckles — APPROVED** [review]

**Comment by kahirokunn** [comment]
How can I get a CI through?

**Comment by gandro** [comment]
> How can I get a CI through?

CI can only be triggered by Cilium committers (see https://docs.cilium.io/en/v1.13/community/governance/commit_access/).  Just ping one of the assigned reviewers to run the tests for you.

File: install/kubernetes/cilium/templates/clustermesh-apiserver/service.yaml [hunk 1]
```diff
@@ -24,4 +24,6 @@ spec:
   {{- if and (eq "LoadBalancer" .Values.clustermesh.apiserver.service.type) .Values.clustermesh.apiserver.service.loadBalancerIP }}
   loadBalancerIP: {{ .Values.clustermesh.apiserver.service.loadBalancerIP }}
   {{- end }}
+  externalTrafficPolicy: {{ .Values.clustermesh.apiserver.service.externalTrafficPolicy }}
+  internalTrafficPolicy: {{ .Values.clustermesh.apiserver.service.internalTrafficPolicy }}
```
Comment (gandro) [inline install/kubernetes/cilium/templates/clustermesh-apiserver/service.yaml]: This field (`internalTrafficPolicy`) is only stable in Kubernetes 1.26, which s very recent and thus this manifest will fail on older installations.

Let's therefore make it optional. I would suggest the following:

- Set the `.Values.clustermesh.apiserver.service.externalTrafficPolicy` and `.Values.clustermesh.apiserver.service.internalTrafficPolicy` to empty strings by default (in the values.yml.tmpl file)
- Gate those two fields here via an if statement as follows:


```suggestion
  {{- if .Values.clustermesh.apiserver.service.externalTrafficPolicy }}
  externalTrafficPolicy: {{ .Values.clustermesh.apiserver.service.externalTrafficPolicy }}
  {{- end }}
  {{- if .Values.clustermesh.apiserver.service.internalTrafficPolicy }}
  internalTrafficPolicy: {{ .Values.clustermesh.apiserver.service.internalTrafficPolicy }}
  {{- end }}
```

**Review by gandro — CHANGES_REQUESTED** [review]
Thanks for the PR! I have left s ome feedback on Kubernetes compatibility.

Also, on a different note, the PR/commit description mentions `ci:` - we use the term `ci` typically for continuous integration. I think a better commit title here would be `clustermesh-apiserver: ExternalTrafficPolicy and internalTrafficPolicy can now be changed`

File: install/kubernetes/cilium/values.yaml.tmpl [hunk 2]
```diff
@@ -2341,6 +2341,12 @@ clustermesh:
       # For EKS LoadBalancer, use annotation service.beta.kubernetes.io/aws-load-balancer-internal: 0.0.0.0/0
       annotations: {}
 
+      # -- The externalTrafficPolicy of service used for apiserver access.
+      externalTrafficPolicy: Cluster
+
+      # -- The externalTrafficPolicy of service used for apiserver access.
+      internalTrafficPolicy: Cluster
```
Comment (qmonnet) [inline install/kubernetes/cilium/values.yaml.tmpl]: The two comments are identical, I guess this one should say `internalTrafficPolicy`?

```suggestion
      # -- The internalTrafficPolicy of service used for apiserver access.
      internalTrafficPolicy: Cluster
```

**Review by qmonnet — CHANGES_REQUESTED** [review]
> ci: ExternalTrafficPolicy and internalTrafficPolicy can now be changed.

It would be nice to add a pointer to the commit that enabled this change, in the commit description.

Please note that the smoke tests for the Helm charts are failing:
```
 please run 'make -C install/kubernetes' and submit your changes
```

File: install/kubernetes/cilium/values.yaml.tmpl (hunk 2 above)
Comment (kahirokunn) [reply install/kubernetes/cilium/values.yaml.tmpl]: thx 🙏

File: install/kubernetes/cilium/templates/clustermesh-apiserver/service.yaml (hunk 1 above)
Comment (kahirokunn) [reply install/kubernetes/cilium/templates/clustermesh-apiserver/service.yaml]: @gandro 
Thx!
Fixed and squashed to one commit.
How do I add your contribution to the commit?

File: install/kubernetes/cilium/values.yaml.tmpl (hunk 2 above)
Comment (kahirokunn) [reply install/kubernetes/cilium/values.yaml.tmpl]: @qmonnet 
Thx!
Fixed and squashed to one commit.
How do I add your contribution to the commit?

**Comment by kahirokunn** [comment]
I am getting an error in CI, but my commit message clearly does not exceed 75 characters.
Is this a bug?

```
make -C bpf checkpatch

=========================================================
[1/14] Running on b323347a6513d262dce69033275572f97dcd04d7
clustermesh-apiserver: ExternalTrafficPolicy and internalTrafficPolicy can now be changed.
=========================================================

"[PATCH] clustermesh-apiserver: ExternalTrafficPolicy and" has no obvious style problems and is ready for submission.

NOTE: Ignored message types: C99_COMMENT_TOLERANCE COMMIT_LOG_LONG_LINE COMMIT_MESSAGE COMPLEX_MACRO CONSTANT_CONVERSION CONST_STRUCT EMAIL_SUBJECT FILE_PATH_CHANGES FROM_SIGN_OFF_MISMATCH GIT_COMMIT_ID JIFFIES_COMPARISON LEADING_SPACE LONG_LINE_COMMENT MACRO_WITH_FLOW_CONTROL MULTISTATEMENT_MACRO_USE_DO_WHILE NOT_UNIFIED_DIFF PRINTK_WITHOUT_KERN_LEVEL TRAILING_SEMICOLON TRAILING_STATEMENTS VOLATILE
ERROR:CUSTOM: Please avoid long commit subjects (max: 75, found: 90)
```

**Comment by gandro** [comment]
> I am getting an error in CI, but my commit message clearly does not exceed 75 characters.

I think it does, currently it's set to
```
clustermesh-apiserver: ExternalTrafficPolicy and internalTrafficPolicy can now be changed.
```
which actually is 90 chars long

**Comment by kahirokunn** [comment]
Thx!

**Comment by kahirokunn** [comment]
How about now?

**Review by gandro — APPROVED** [review]
Thanks!

**Review by qmonnet — APPROVED** [review]
Thanks

**Comment by kahirokunn** [comment]
Thanks too! 🙏

**Comment by kahirokunn** [comment]
I think the CI test is failing in some areas that have nothing to do with my fix.
Does this happen often?

**Comment by gandro** [comment]
> I think the CI test is failing in some areas that have nothing to do with my fix. Does this happen often?

Unfortunately, the push to the branch invalidated the results so I cannot check them anymore. But yes, there are a few flaky tests.

Let me re-run CI to check if the test failures are legitimate or not. Please do not push to the branch anymore (unless to fix bugs). Rebasing is usually not needed.

**Comment by kahirokunn** [comment]
okay

**Comment by gandro** [comment]
CI passed - the conformance ingress tests [have been refactored](https://github.com/cilium/cilium/pull/24025) and so will not be triggered in this branch. Old old test passed here: https://github.com/cilium/cilium/actions/runs/4350648456/jobs/7601557102

Marking ready to merge. Thanks @kahirokunn for your contribution!

**Pull request merged**